storage/flush.go

```diff
@@ -21,157 +21,132 @@
-		opts:           opts,
-		fs:             flushNotStarted,
-		flushAttempted: make(map[time.Time]flushState),
-		flushTimes:     make([]time.Time, 0, numSlots),
+		database:    database,
+		opts:        opts,
+		blockSize:   blockSize,
+		pm:          pm,
+		flushStates: map[time.Time]fileOpState{},
 	}
 }
 
-func (fm *flushManager) NeedsFlush(t time.Time) bool {
-	// If we haven't bootstrapped yet, don't flush.
-	if !fm.database.IsBootstrapped() {
-		return false
-	}
+func (mgr *flushManager) HasFlushed(t time.Time) bool {
+	mgr.RLock()
+	defer mgr.RUnlock()
 
-	firstBlockStart := fm.getFirstBlockStart(t)
-	fm.RLock()
-	defer fm.RUnlock()
-	// If we are in the middle of flushing data, don't flush.
-	if fm.fs == flushInProgress {
+	flushState, exists := mgr.flushStates[t]
+	if !exists {
 		return false
 	}
-	// If we have already tried flushing for this block start time, don't try again.
-	if _, exists := fm.flushAttempted[firstBlockStart]; exists {
-		return false
-	}
-	return true
+	return flushState.Status == fileOpSuccess
 }
 
-func (fm *flushManager) getFirstBlockStart(t time.Time) time.Time {
-	rops := fm.opts.GetRetentionOptions()
-	bufferPast := rops.GetBufferPast()
-	blockSize := rops.GetBlockSize()
-	return t.Add(-bufferPast).Add(-blockSize).Truncate(blockSize)
+func (mgr *flushManager) FlushTimeStart(t time.Time) time.Time {
+	retentionPeriod := mgr.opts.GetRetentionOptions().GetRetentionPeriod()
+	return t.Add(-retentionPeriod).Truncate(mgr.blockSize)
 }
 
-func (fm *flushManager) Flush(t time.Time, async bool) {
-	timesToFlush := fm.getTimesToFlush(t)
-	if len(timesToFlush) == 0 {
-		return
-	}
+func (mgr *flushManager) FlushTimeEnd(t time.Time) time.Time {
+	bufferPast := mgr.opts.GetRetentionOptions().GetBufferPast()
+	return t.Add(-bufferPast).Add(-mgr.blockSize).Truncate(mgr.blockSize)
+}
 
-	fm.Lock()
-	if fm.fs == flushInProgress {
-		fm.Unlock()
-		return
+func (mgr *flushManager) Flush(t time.Time) error {
+	timesToFlush := mgr.flushTimes(t)
+	if len(timesToFlush) == 0 {
+		return nil
 	}
-	fm.fs = flushInProgress
-	fm.Unlock()
-
-	flushFn := func() {
-		ctx := fm.opts.GetContextPool().Get()
-		defer ctx.Close()
-
-		for _, flushTime := range timesToFlush {
-			success := fm.flushWithTime(ctx, flushTime)
-			fm.Lock()
-			flushState := fm.flushAttempted[flushTime]
-			if success {
-				flushState.Status = flushSuccess
-			} else {
-				flushState.Status = flushFailed
-				flushState.NumFailures++
-			}
-			fm.flushAttempted[flushTime] = flushState
-			fm.Unlock()
+	ctx := mgr.opts.GetContextPool().Get()
+	defer ctx.Close()
+
+	multiErr := xerrors.NewMultiError()
+	for _, flushTime := range timesToFlush {
+		mgr.Lock()
+		if !mgr.needsFlush(flushTime) {
+			continue
 		}
-
-		fm.Lock()
-		fm.fs = flushNotStarted
-		fm.Unlock()
+		flushState := mgr.flushStates[flushTime]
+		flushState.Status = fileOpInProgress
+		mgr.flushStates[flushTime] = flushState
+		mgr.Unlock()
+
+		flushErr := mgr.flushWithTime(ctx, flushTime)
```

Comment:
It's because `flushWithTime` could take a long time to finish while writing the in-memory data out to disk so we don't want to hold the lock while data is being flushed.